File: WorkflowService/Workflow.DAL/Models/Goal.cs
using System;
using System.Collections.Generic;

namespace Workflow.DAL.Models
{
    public class Goal
    {
        public int Id { get; set; }

        public int GoalNumber { get; set; }
        public string Title { get; set; }
        public string Description { get; set; }

        public int ProjectId { get; set; }
        public Project Project { get; set; }

        public int? ParentGoalId { get; set; }
        public Goal ParentGoal { get; set; }
        public List<Goal> ChildGoals { get; set; } = new List<Goal>();

        public DateTime CreationDate { get; set; }
        public DateTime ExpectedCompletedDate { get; set; }
        public TimeSpan EstimatedPerformingTime { get; set; }

        public GoalState State { get; set; } = GoalState.New;
        public GoalPriority Priority { get; set; } = GoalPriority.Normal;

        public string OwnerId { get; set; }
        public ApplicationUser Owner { get; set; }

        public string PerformerId { get; set; }
        public ApplicationUser Performer { get; set; }

        public List<GoalObserver> Observers { get; set; } = new List<GoalObserver>();

        public List<Attachment> Attachments { get; set; } = new List<Attachment>();

        public bool IsRemoved { get; set; }
    }

    /// <summary>
    /// Состояние выполнения задачи
    /// </summary>
    public enum GoalState
    {
        /// <summary>
        /// Новая
        /// </summary>
        New,

        /// <summary>
        /// Выполняется
        /// </summary>
        Perform,

        /// <summary>
        /// Приостановлена
        /// </summary>
        Delay,

        /// <summary>
        /// Тестируется
        /// </summary>
        Testing,
        
        /// <summary>
        /// Прошла тестирование и успешно завершена
        /// </summary>
        Succeed,

        /// <summary>
        /// Отклонена по результатам тестирования
        /// </summary>
        Rejected
    }

    /// <summary>
    /// Приоритет задачи
    /// </summary>
    public enum GoalPriority
    {
        /// <summary>
        /// Низкий
        /// </summary>
        Low,

        /// <summary>
        /// Нормальный 
        /// </summary>
        Normal,

        /// <summary>
        /// Высокий
        /// </summary>
        High
    }
}

File: WorkflowClient/src/core/entities/goal/goal.ts
import moment from 'moment'
import { IGoal, IGoalData, Priority, Status } from './goal.types'
import Attachment from '@/modules/goals/models/attachment.type'

export class Goal implements IGoal {
  id?: number
  goalNumber?: number
  title?: string
  description?: string
  projectId?: number
  projectName?: string
  parentGoalId?: number
  ownerId?: string
  ownerFio?: string
  performerId?: string
  performerFio?: string
  creationDate?: string
  expectedCompletedDate?: string
  estimatedPerformingTime?: string
  state: Status = Status.New
  priority?: Priority
  hasChildren?: boolean
  isAttachmentsExist?: boolean
  isRemoved?: boolean

  children?: IGoal[]
  attachments?: Attachment[]

  constructor(data: IGoalData) {
    if (data.id) this.id = data.id
    this.goalNumber = data.goalNumber
    this.title = data.title
    this.description = data.description
    this.projectId = data.projectId
    this.projectId = data.projectId
    this.parentGoalId = data.parentGoalId
    this.ownerId = data.ownerId
    this.ownerFio = data.ownerFio
    this.performerId = data.performerId
    this.performerFio = data.performerFio
    if (data.creationDate) this.creationDate = data.creationDate
    this.expectedCompletedDate = data.expectedCompletedDate
    this.estimatedPerformingTime = data.estimatedPerformingTime
    this.state = data.state
    this.priority = data.priority
    this.hasChildren = data.hasChildren
    this.isAttachmentsExist = data.isAttachmentsExist
    this.isRemoved = data.isRemoved
  }
}
